File: src/algorithms/searching/binary/sqrt-binary-search/sources/sqrt-binary-search.py
def sqrt_binary_search(target_value: int) -> int:  # @step:initialize
    if target_value < 2:  # @step:initialize
        return target_value  # @step:initialize
    low_index = 1  # @step:initialize
    high_index = target_value // 2  # @step:initialize
    result_index = 0  # @step:initialize

    while low_index <= high_index:
        mid_index = (low_index + high_index) // 2  # @step:compare
        mid_squared = mid_index * mid_index  # @step:compare

        if mid_squared == target_value:  # @step:compare,found
            return mid_index  # @step:found
        elif mid_squared < target_value:  # @step:eliminate
            # mid_index is a candidate floor — search for a larger value
            result_index = mid_index  # @step:eliminate
            low_index = mid_index + 1  # @step:eliminate
        else:  # @step:eliminate
            # mid_index is too large — search left
            high_index = mid_index - 1  # @step:eliminate

    return result_index  # @step:complete

File: src/algorithms/searching/binary/sqrt-binary-search/sources/sqrt-binary-search.ts
// Square Root via Binary Search — find the integer square root of a non-negative number
function sqrtBinarySearch(targetValue: number): number {
  // @step:initialize
  if (targetValue < 2) return targetValue; // @step:initialize
  let lowIndex = 1; // @step:initialize
  let highIndex = Math.floor(targetValue / 2); // @step:initialize
  let resultIndex = 0; // @step:initialize

  while (lowIndex <= highIndex) {
    const midIndex = Math.floor((lowIndex + highIndex) / 2); // @step:compare
    const midSquared = midIndex * midIndex; // @step:compare

    if (midSquared === targetValue) {
      // @step:compare,found
      return midIndex; // @step:found
    } else if (midSquared < targetValue) {
      // @step:eliminate
      // midIndex is a candidate floor — search for a larger value
      resultIndex = midIndex; // @step:eliminate
      lowIndex = midIndex + 1; // @step:eliminate
    } else {
      // @step:eliminate
      // midIndex is too large — search left
      highIndex = midIndex - 1; // @step:eliminate
    }
  }

  return resultIndex; // @step:complete
}
